### PR TITLE
feat: Use ranges instead of a vector for input witness

### LINKED ISSUE #861: Change ABI to Witness map to work on ranges.
# Problem

The size of the ABI file currently scales with the number of field elements which make up the circuit inputs/outputs.

As we scale up the number of fields (e.g. for a circuit which takes an image as an input) we're going to get an unreasonably large ABI.

# Solution

We encode all inputs to contiguous ranges of witness indices so we can change `param_witnesses` to be a `BTreeMap<String, Range<Witness>>` to handle this.

The only trouble with this is that the public outputs may not be in order (e.g. say the return value is an array where one of the elements is also a public input and so has a smaller index) so this breaks.

We could then use a `BTreeMap<String, Vec<Range<Witness>>>` to handle all of these cases or handle the return value specially (which may be preferable considering #678)

# Alternatives considered

# Additional context

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #861

## Summary\*

Use ranges for representing witness inputs, instead of using a vector.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs [hunk 1]
```diff
@@ -239,8 +240,11 @@ impl Context {
         for acir_var in output_vars {
             self.acir_context.return_var(acir_var)?;
         }
-
-        Ok(self.acir_context.finish(witness_inputs, Vec::new()))
+        let mut wit = Vec::new();
+        for input in witness_inputs {
+            wit.push(RangeInclusive::new(input, input));
+        }
+        Ok(self.acir_context.finish(wit, Vec::new()))
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:247]:
```suggestion
        let witnesses = vecmap(witness_inputs, |input| RangeInclusive::new(input, input));
        Ok(self.acir_context.finish(witnesses, Vec::new()))
```

File: tooling/noirc_abi/src/lib.rs [hunk 2]
```diff
@@ -529,6 +530,18 @@ impl ContractEvent {
...
 }
 
+fn range_to_vec(ranges: &Vec<Range<Witness>>) -> Vec<Witness> {
+    let mut result = Vec::new();
+    for range in ranges {
+        let mut w = range.start.witness_index();
+        while w != range.end.witness_index() {
+            result.push(w.into());
+            w += 1;
+        }
```
Comment (jfecher) [inline tooling/noirc_abi/src/lib.rs:540]:
```suggestion
        for witness in range.start.witness_index() .. range.end.witness_index() {
            result.push(w.into());
        }
```

File: tooling/noirc_abi/src/lib.rs [hunk 3]
```diff
@@ -529,6 +530,18 @@ impl ContractEvent {
     }
 }
 
+fn range_to_vec(ranges: &Vec<Range<Witness>>) -> Vec<Witness> {
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/lib.rs:533]:
```suggestion
fn range_to_vec(ranges: &[Range<Witness>]) -> Vec<Witness> {
```

**Comment by TomAFrench** [comment]
noir_js tests need to be updated

File: tooling/noirc_abi/src/lib.rs (hunk 3 above)
Comment (guipublic) [reply tooling/noirc_abi/src/lib.rs]: Done

File: tooling/noirc_abi/src/lib.rs (hunk 2 above)
Comment (guipublic) [reply tooling/noirc_abi/src/lib.rs]: Done

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs (hunk 1 above)
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: Done

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**